Source/Dafny/AST/DafnyAst.cs

```diff
@@ -6927,8 +6927,35 @@ public Statement(IToken tok, IToken endTok)
 
     /// <summary>
     /// Returns the non-null expressions of this statement proper (that is, do not include the expressions of substatements).
+    /// Filters all sub expressions that are not part of specifications
     /// </summary>
-    public virtual IEnumerable<Expression> SubExpressions {
+    public IEnumerable<Expression> SubExpressions {
+      get {
+        foreach (var e in SpecificationSubExpressions) {
+          yield return e;
+        }
+
+        foreach (var e in NonSpecificationSubExpressions) {
+          yield return e;
+        }
+      }
+    }
+
+    /// <summary>
+    /// Returns the non-null expressions of this statement proper (that is, do not include the expressions of substatements).
+    /// Filters only expressions that are always part of specifications
+    /// </summary>
+    public virtual IEnumerable<Expression> SpecificationSubExpressions {
```

Comment:
Did you consider using the term `Ghost` instead of `Specification` here? What are the pros/cons?

Reply:
Indeed, I could have detailed my thinking to double-proof it with the rest of the team.

Pros of using the wording "ghost"
* Seems consistent with the rest, "ghost" meaning "non-compiled"
* Shorter than "specification"

Cons of using the wording "ghost"
* I wanted a clear disjunction between the two types of expressions, so I had the choice between "Ghost/Compiled" or "Specification/NonSpecification". The problem with "Compiled" (or "NonGhost"?) is that such expressions can still be ghost if they are in a ghost context (like a lemma or a ghost method). Ghost seems like a final property, where the code appears in the final compiler or not. Here I want the reasoning about the expressions to be local.
* We currently mark things as "Ghost" in the IDE only when they are ambiguously ghost (I would have preferred to visualize everything ghost the same way, but that's the current consensus). Here, we are dealing with expressions that are unambiguously "ghost", so they are specifications, like pre and post conditions.

Reply:
Thanks, for explaining. Makes sense :-)